include/onnxruntime/core/platform/EigenNonBlockingThreadPool.h

```diff
@@ -902,124 +842,265 @@ void EndParallelSectionInternal(PerThread &pt,
   // Notify workers to exit from the section
   ps.active = false;
 
-  if (ps.dispatch_q_idx > -1) {
+  // First, attempt to revoke the dispatch task.  If we succeed then
+  // we know we revoked _something_ pushed for the current loop.  That
+  // may be the dispatch task itself, or it may be a task pushed by
+  // the dispatch task.  Those cases are distinguished by whether or
+  // not the dispatch task itself has started -- if it has not started
+  // then it cannot have pushed tasks.
+  if (ps.dispatch_q_idx != -1) {
     Queue& q = worker_data_[ps.dispatch_q_idx].queue;
     if (q.RevokeWithTag(pt.tag, ps.dispatch_w_idx)) {
-      ps.dispatch_q_idx = -1; // cancel dispatch if not started yet
-    } else {
-      // if dispatch task started, wait for its dispatch completion
-      while (!ps.dispatch_done.load(std::memory_order_acquire)) {
-        onnxruntime::concurrency::SpinPause();
+      if (!ps.dispatch_started.load(std::memory_order_acquire)) {
```

Comment:
>dispatch_started [](http://example.com/codeflow?start=14&length=16)

If revoke succeed, shouldn't "dispatch_started" be false for sure?

Reply:
There was an interesting corner case here.  I will extend the comment to capture it more clearly.  Suppose we have a loop with DoP=4. 
 What could happen is:

* Main thread schedules dispatch task D1
* Dispatch task dispatches additional tasks T1, T2
* Main thread completes all of the work in the loop
* Main thread attempts to end the loop
* Main thread attempts to revoke dispatcher task D1
* Revoke succeeds, but it actually revokes T1 (which may have been placed in the exact same slot in the queue that D1 used)
* Loop end gets stuck waiting for 2 tasks